### ISSUE TITLE
Some DX improvements to Route?

### ISSUE THREAD

**Issue by wmertens** [ReactTraining/react-router#4548]
(I'm going through my app implementing v4 and trying to feel where the API feels a bit rough)

I'm not a fan of function children. Children are normally React elements, so that's breaking at least the implicit React API, and there are infinite other attributes that can be used to pass functions.

Furthermore, preventing during-render creation of functions is always a plus.

Therefore, I would like to propose these two additions to `Route`, completely backwards compatible:

1. **always**: `<Route path="/maybeMatch" always render={...}/>`
  The `always` attribute would also call `render` if the path doesn't match, like `children`. The word `render` instantly conveys the proper meaning, whereas `children` does not.
2. **props**: `<Route component={Foo} props={bar}/>`
  This is equivalent to `<Route render={args => <Foo {...args} {...bar}/>}/>`, except more readable as well as preventing the creation of an extra function at render time.

Thoughts?

And if I might go one step further, an almost-breaking change would be to only render `children` on match if `children` is not a function. This would make it behave less surprisingly, especially when combined with relative paths:

```js
<Route component={SomeWrapper}>
  <div>
    This will always render, as before, because there is no path on the parent
    <Route path="whatever" .../>
    <Route .../>
  </div>
</Route>
<Route path="/onlyHere" component={SomeWrapper}>
  <div>
    This would only render for urls starting with /onlyHere, not like before
    <Route path="whatever" .../>
    <Route .../>
  </div>
</Route>
```

This would actually make it behave more like the `routes` configuration of previous RR versions.

So in summary: **always**, **props**, and not rendering regular children if there is no match.

**Comment by timdorr** [comment]
A React component can be just a function now ;)

Also, this is the main reason for child functions https://gist.github.com/ryanflorence/a301dc184f75e929a263dc1e80399a28

**Comment by wmertens** [comment]
Sure I understand why you would want to thunk the rendering. However, using `children` for this feels "dirty", even though it works.

**Comment by pshrmn** [comment]
I have thought before how it would be nice to have something like the `props` prop that you mention (although I thought of it more as `inject`, but figuring out the "best" name is just bikeshedding). I sort of go back and forth on whether or not I consider it a good idea.

```js
// an added prop to pass props to the component when it is rendered
<Route path='/form-page' component={Form} inject={prefilledData} />
// can be done right now using the render prop
<Route path='/form-page' render={() => <Form {...prefilledData} />} />
// and because the function passed to render is a stateless component,
// it also works with the component prop
<Route path='/form-page' component={() => <Form {...prefilledData} />} />
```
Obviously you are creating a new function on every re-render, but I'm not really sure what the performance implications are.

I don't think that the `always` prop is necessary. Maybe it would be clearer if examples used the implicit `children`? Then it looks more like regular JSX.

```js
<Route path='/some-place'>
  {({ match }) => <div>Does my path match the current location? { match ? 'Yes' : 'No' }</div>}
</Route>
```
With the addition of the `<Switch>`, I think that the major use case of the `children` prop is in composition, not for actually matching routes (see the `<NavLink>`).

**Comment by wmertens** [comment]
Well I wanted to see the impact and created PR #4550 - as you can see, very simple changes.

**Comment by pshrmn** [comment]
The code is simple, but it is also adding duplicate functionality.

I think that the only reason that `always` would be necessary is if support for `children` as a function was removed. I could see an argument being made for that, but for the time being I don't think adding a second way to always match is that beneficial.

**Comment by wmertens** [comment]
@pshrmn this is an example of `always` being useful: a `MenuItem` component that renders "active" when `match` is truthy:

 ```js
  <Route path="/one" always component={MenuItem} props={{icon: "star"}}/>
  ```

Without `always`, this *has* to be a `children` function, which is a stylistic choice that I believe RR should not take sides in.

```js
<Route path="/one">
  {({match}) => <MenuItem icon="star" match={match}/>}
</Route>
```

**Comment by mjackson** [comment]
We've gotta be careful when adding API purely for convenience, because the bikeshedding can go on forever. The important thing at this point is we already give you all the API you need to do whatever you want.

FWIW, I'd personally write your last example like this:

```js
<Route path="/one" children={({match}) => (
  <MenuItem icon="star" match={match}/>
)}/>
```

So I'm not convinced that we need an `always` prop. `children` already does that. As for `render` being a better name than `children`, I agree that it's more descriptive. But it may help to think of things this way:

`children` is the original prop React uses for rendering. In other words, when a component wants to know the answer to "what should I render?" it uses `this.props.children`. So that's the prop we built our rendering API around. It's the *original* "render prop".

**Both `component` and `render` are already convenience (or "sugar") API**

`component` is a prop that is designed to replace the common case of

```js
<Route children={({ match }) => (
  match ? <Something/> : null
)}/>
```

with

```js
<Route component={Something}/>
```

because rendering a component when a route matches is a super common use case.

Adding a `props` prop to make sure we can use `component` to do the exact same thing as the API it was designed to replace feels redundant. It's going backwards. It's saying, essentially, "we gave you a `component` prop, but then realized that we took some level of control away from you, so we had to give you a `props` prop as well to make up for the stuff we took away."

`render` is a prop that we give you for the common case of rendering something when you match. This one is slightly more powerful than `component` because you can pass in whatever other props you want. In other words, instead of writing

```js
<Route children={({ match }) => (
  match ? <Something some="other" stuff/> : null
)}/>
```

you can now write

```js
<Route render={() => (
  <Something some="other" stuff/>
)}/>
```

Again, adding more sugar (`always`) to make the sugar API behave exactly as the API it was designed to make more convenient (`children` with a `match`) feels redundant.

If anything, I'd just remove the `component` and `render` APIs before adding more sugar to make up for functionality you don't have when you use them instead of just using `children`.

**Comment by wmertens** [comment]
@mjackson Ah well, worth a try. It grates a bit to have a
function-as-children, but I'll just hum a bit while working Routes using
them.

One last thing though, how do you feel about regular children only being
rendered on match? As I explained, this makes a Route config behave more
like before.

**Comment by wmertens** [comment]
@mjackson not sure if you saw the question I added to the comment above: how do you feel about regular children only being rendered on match? As I explained, this makes a Route config behave more like before. I think this will be important once relative routes land.

Specifically, https://github.com/ReactTraining/react-router/pull/4550/commits/eff2d6175198cc709825f3735f84f109a18f4bc6
